pybossa/api/bulktasks.py

```diff
@@ -0,0 +1,95 @@
+# -*- coding: utf8 -*-
+# This file is part of PYBOSSA.
+#
+# Copyright (C) 2022 Scifabric LTD.
+#
+# PYBOSSA is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# PYBOSSA is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with PYBOSSA.  If not, see <http://www.gnu.org/licenses/>.
+"""
+PYBOSSA api module for exposing domain object BulkTaskUpdate via an API.
+"""
+from flask import request, Response
+from flask import current_app
+from flask_login import current_user
+from werkzeug.exceptions import NotFound, BadRequest
+from werkzeug.exceptions import MethodNotAllowed, Unauthorized
+from pybossa.cache import memoize
+from pybossa.core import project_repo, task_repo, timeouts
+from pybossa.error import ErrorStatus
+from pybossa.api.task import TaskAPI
+from pybossa.model.task import Task
+from pybossa.util import jsonpify, crossdomain
+from pybossa.core import ratelimits
+from pybossa.ratelimit import ratelimit
+from pybossa.util import admin_or_subadmin_required
+
+
+cors_headers = ["Content-Type", "Authorization"]
+allowed_attributes = ["id", "priority_0"]
+error = ErrorStatus()
+
+class BulkTasksAPI(TaskAPI):
+
+    """Class for domain object Task."""
+
+    __class__ = Task
+
+    def get(self, oid=None):
+        raise MethodNotAllowed
+
+    def post(self):
+        raise MethodNotAllowed
+
+    def delete(self, oid=None):
+        raise MethodNotAllowed
+
+    @jsonpify
+    @crossdomain(origin="*", headers=cors_headers)
+    @ratelimit(limit=ratelimits.get("LIMIT"), per=ratelimits.get("PER"))
+    def put(self, oid):
+        """Update task attributes in bulk. Need atleast subadmin access"""
+        if current_user.is_anonymous or not(current_user.admin or current_user.subadmin):
+            raise Unauthorized("Insufficient privilege to the request")
+
+        project = project_repo.get(oid)
+        update_info = request.json
+        if not (project and update_info):
+            raise BadRequest
+
+        all_owners = project.owners_ids + ([project.owner_id] if project.owner_id not in project.owners_ids else [])
+        if not (current_user.admin or current_user.id in all_owners):
```

Comment:
This is custom auth check based on project id passed in which is used for determining 1. users api key has access to this project & 2. the tasks belonging to this project. With this we avoid having auth access check for each task. It seems `ensure_auth` check for each task can be avoided with this custom check.

Reply:
line 67-69 checks user access on project level instead of task level, so "2. tasks belong to this project" is not checked here. We can replace this code with `ensure_auth` to confirm that user api key has WRITE access to this project. Does this make sense?

Reply:
Do you mean replacing line 67-69 with
```python
ensure_authorized_to("update", project)
```
this would work as well, just that it somehow doesn't seems logically correct as we're not updating project's settings/configuration. What do you think?

Reply:
You make a good point. Technically we should validate WRITE access to task instead of project.

However, the requirement of WRITE task and project are same, they both need admin or sub admin_coowner. https://github.com/bloomberg/pybossa/blob/1e94133f0468c2956f9fdc1efe6317bd92a589c8/pybossa/auth/project.py#L28

I do not have a strong preference here. you can make the decision.